### PR TITLE
system: handle PurePath

### PR DESCRIPTION
**Pull request by efiop**
```
------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/efiop/git/dvc/dvc/remote/local/__init__.py", line 280, in upload
    no_progress_bar=no_progress_bar,
  File "/Users/efiop/git/dvc/dvc/utils/__init__.py", line 129, in copyfile
    System.reflink(src, dest)
  File "/Users/efiop/git/dvc/dvc/system.py", line 125, in reflink
    ret = System._reflink_darwin(source, link_name)
  File "/Users/efiop/git/dvc/dvc/system.py", line 89, in _reflink_darwin
    ctypes.c_char_p(src.encode("utf-8")),
AttributeError: 'PosixPathInfo' object has no attribute 'encode'
------------------------------------------------------------
```
Signed-off-by: Ruslan Kuprieiev <kupruser@gmail.com>

* [x] Have you followed the guidelines in our
      [Contributing document](https://dvc.org/doc/user-guide/contributing)?

* [x] Does your PR affect documented changes or does it add new functionality
      that should be documented? If yes, have you created a PR for
      [dvc.org](https://github.com/iterative/dvc.org) documenting it or at
      least opened an issue for it? If so, please add a link to it.

-----

### CODE REVIEW

**Comment by Suor** [comment]
You should use `fspath()` not `str()` there. Maybe one just should not pass paths there though, that's up to where you want to handle that. BTW, you don't need to check object type before passing to `fspath()`.

File: dvc/system.py [hunk 1]
```diff
@@ -85,6 +85,12 @@ def _reflink_darwin(src, dst):
         clonefile.argtypes = [ctypes.c_char_p, ctypes.c_char_p, ctypes.c_int]
         clonefile.restype = ctypes.c_int
 
+        if isinstance(src, pathlib.PurePath):
+            src = str(src)
+
+        if isinstance(dst, pathlib.PurePath):
+            dst = str(dst)
```
Comment (Suor) [inline dvc/system.py]:
```python
src, dst = fspath(src), fspath(dst)
```

**Comment by efiop** [comment]
@Suor Thanks for the review. Adjusted. Please take a look.

**Comment by efiop** [comment]
Ah, just a sec... Will adjust it differently .

**Comment by efiop** [comment]
Ok, so we shouldn't have been converting with fspath() too early, but rather make System.* methods support PurePath-like objects. @Suor Take a look.

**Pull request merged**